File: solutions/javascript/JumpGameII/JumpGameII.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var jump = function(nums) {
    let res = 0, max = 0, cur_p = 0;
    for (let i = 0; i < nums.length - 1; i++) {
        max = Math.max(max, i + nums[i]);
        if (i === cur_p) {
            res++;
            cur_p = max;
        }
    }
    return res;
};

File: solutions/java/JumpGameII/JumpGameII.java
class Solution {
    public int jump(int[] nums) {
        //BFS
        if (nums.length < 2) {
            return 0;
        }
        int level = 0, curFarthest = 0, curEnd = 0;
        for(int i = 0; i < nums.length; i++) {
            curFarthest = Math.max(curFarthest, nums[i] + i);
            if(i == curEnd) {
                curEnd = curFarthest;
                level++;
            }
            if(curEnd >= nums.length - 1) {
                return level;
            }
        }
        return -1;
    }
}